arangod/Aql/types.h

```diff
@@ -60,9 +60,121 @@ struct Collection;
 /// @brief type for variable ids
 typedef uint32_t VariableId;
 
+typedef uint16_t RegisterCount;
+
 /// @brief type for register numbers/ids
-typedef unsigned RegisterId;
-typedef size_t RegisterCount;
+struct RegisterId {
+  enum class Type : uint16_t {
+    Regular = 0,
+    Const = 1
+  };
+
+  using value_t = uint16_t;
+  static_assert(std::is_same_v<value_t, RegisterCount>);
+
+  constexpr RegisterId() = default;
+
+  static constexpr value_t maxRegisterId = 1000;
+  
+  // TODO - make constructor explicit (requires a lot of changes in test code)
+  constexpr RegisterId(value_t v, Type type = Type::Regular)
+      : _value(v), _type(type) {
+    TRI_ASSERT(v <= maxRegisterId);
+  }
+
+  constexpr static RegisterId makeConst(value_t value) noexcept {
+    return RegisterId(value, Type::Const);
+  }
+
+  constexpr static RegisterId makeRegular(value_t value) noexcept {
+    return RegisterId(value, Type::Regular);
+  }
+
+  constexpr static RegisterId fromUInt32(uint32_t value) noexcept {
+    auto v = static_cast<value_t>(value);
+    auto type = static_cast<Type>(value >> (sizeof(value_t) * 8));
```

Comment:
what is the behavior if an invalid input value is passed into this function, e.g. UINT32_MAX?

Reply:
Then `isValid()` would be false. At the moment there is only an assertion, would you prefer to throw an exception if we get some invalid value here?

Reply:
I would be happier if it wasn't possible to create invalid `RegisterId` objects at all.

Reply:
Allowing to create an object with invalid state and later having to check if it is valid feels a bit like in the 80s.

Reply:
Done!